x-pack/plugins/ml/public/application/components/data_grid/data_grid.tsx

```diff
@@ -50,132 +51,152 @@ function isWithHeader(arg: any): arg is PropsWithHeader {
-        gridStyle={euiDataGridStyle}
-        rowCount={rowCount}
-        renderCellValue={renderCellValue}
-        sorting={{ columns: sortingColumns, onSort }}
-        toolbarVisibility={euiDataGridToolbarSettings}
-        pagination={{
-          ...pagination,
-          pageSizeOptions: [5, 10, 25],
-          onChangeItemsPerPage,
-          onChangePage,
-        }}
-      />
-    </div>
-  );
-};
+      );
+    }
+
+    return (
+      <div data-test-subj={`${dataTestSubj} ${status === INDEX_STATUS.ERROR ? 'error' : 'loaded'}`}>
+        {isWithHeader(props) && (
+          <EuiFlexGroup alignItems="center" justifyContent="spaceBetween">
+            <EuiFlexItem>
+              <DataGridTitle title={props.title} />
+            </EuiFlexItem>
+            <EuiFlexItem grow={false}>
+              <EuiCopy
+                beforeMessage={props.copyToClipboardDescription}
+                textToCopy={props.copyToClipboard}
+              >
+                {(copy: () => void) => (
+                  <EuiButtonIcon
+                    onClick={copy}
+                    iconType="copyClipboard"
+                    aria-label={props.copyToClipboardDescription}
+                  />
+                )}
+              </EuiCopy>
+            </EuiFlexItem>
+          </EuiFlexGroup>
+        )}
+        {status === INDEX_STATUS.ERROR && (
+          <div data-test-subj={`${dataTestSubj} error`}>
+            <EuiCallOut
+              title={i18n.translate('xpack.ml.dataGrid.indexDataError', {
+                defaultMessage: 'An error occurred loading the index data.',
+              })}
+              color="danger"
+              iconType="cross"
+            >
+              <EuiCodeBlock language="json" fontSize="s" paddingSize="s" isCopyable>
+                {errorMessage}
+              </EuiCodeBlock>
+            </EuiCallOut>
+            <EuiSpacer size="m" />
+          </div>
+        )}
+        <EuiDataGrid
+          aria-label={isWithHeader(props) ? props.title : ''}
+          columns={columns}
+          columnVisibility={{ visibleColumns, setVisibleColumns }}
+          gridStyle={euiDataGridStyle}
+          rowCount={rowCount}
+          renderCellValue={renderCellValue}
+          sorting={{ columns: sortingColumns, onSort }}
+          toolbarVisibility={euiDataGridToolbarSettings}
+          pagination={{
+            ...pagination,
+            pageSizeOptions: [5, 10, 25],
+            onChangeItemsPerPage,
+            onChangePage,
+          }}
+        />
+      </div>
+    );
+  },
+  (prevProps, nextProps) => isEqual(pickProps(prevProps), pickProps(nextProps))
+);
+
+function pickProps(props: Props) {
+  return [
+    props.columns,
+    props.dataTestSubj,
+    props.errorMessage,
+    props.invalidSortingColumnns,
+    props.noDataMessage,
+    props.pagination,
+    props.rowCount,
+    props.sortingColumns,
+    props.status,
+    props.tableItems,
+    props.toastNotifications,
```

Comment:
not sure you `toastNotifications` should be included in the comparison

Reply:
Good find, removed in 4ac4965!